### PR TITLE
V0.6.0 rc quadtree

### PR DESCRIPTION
**Pull request by khawajasim**
# pyCSEP Pull Request Checklist

Please check out the [contributing guidelines](https://github.com/SCECcode/pycsep/blob/master/CONTRIBUTING.md) for some tips 
on making pull requests to pyCSEP. 

Fixes issue #(*please fill in or delete if not needed*).

## Type of change:

Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation update (this pull request adds no new code)
- [ ] Unpublished science feature (This may require a science review)
- [ ] This change requires a documentation update

## Checklist:

- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes

### CODE REVIEW

File: csep/core/regions.py [hunk 1]
```diff
@@ -1331,3 +1342,313 @@ def california_quadtree_region(magnitudes=None, name="california-quadtree"):
     california_region = QuadtreeGrid2D.from_quadkeys(qk, magnitudes=magnitudes, name=name)
     return california_region
 
+#--------------- Forecast mapping from one grid to another ----------
+def geographical_area_from_qk(quadk):
+    """
+    Wrapper around function geographical_area_from_bounds
+    """
```
Comment (wsavran) [inline csep/core/regions.py:1348]: these should follow the docstring format in the rest of the package

File: csep/core/regions.py [hunk 2]
```diff
@@ -1331,3 +1342,313 @@ def california_quadtree_region(magnitudes=None, name="california-quadtree"):
...
+    ----------
+    quad_cell_id : Stirng
+        Quad key of a cell.
+
+    Returns
+    -------
+    bounds : Mercantile object
+        Latitude and Longitude of bottom left AND top right corners.
+
+    """
```
Comment (wsavran) [inline csep/core/regions.py:1366]: same here

File: csep/core/regions.py [hunk 3]
```diff
@@ -1331,3 +1342,313 @@ def california_quadtree_region(magnitudes=None, name="california-quadtree"):
...
+    # t = Tile.from_quad_tree('{}'.format(quad_cell_id)) # tile from a quad tree repr string #This will change for Mercantile
+    # bounds = t.bounds
```
Comment (wsavran) [inline csep/core/regions.py:1368]: remove commented code if not needed

File: csep/core/regions.py [hunk 4]
```diff
@@ -1331,3 +1341,313 @@ def california_quadtree_region(magnitudes=None, name="california-quadtree"):
...
+    """
+    Required for parallel processing
+    """
+    return geographical_area_from_bounds(cell[0], cell[1], cell[2], cell[3])
+
+
+def _map_overlapping_cells(fcst_grid_poly, fcst_cell_area, fcst_rate_poly, target_poly):  # ,
+    """
+    ------Computationally Expensive --- To be used only for Cells that do not directly conside with target poly -----
+    Note: ALERT !!! This function uses 3 global variables, i.e. fcst_grid_poly, fcst_cell_area, fcst_rate_poly
```
Comment (wsavran) [inline csep/core/regions.py:1390]: how are these global variables? do they need to be defined as `global` in the calling script?

Comment (wsavran) [inline csep/core/regions.py:1390]: be aware of docstring formats. see other functions in the package.

Comment (khawajasim) [reply csep/core/regions.py:1390]: These functions are updating them inside wrapper function. It should not be a problem outside this funciton.

Comment (wsavran) [reply csep/core/regions.py:1390]: so they aren't global but defined within the scope of the calling function as a closure?

**Comment by wsavran** [comment]
@khawajasim im going to close this pr for now for a couple of reasons. first, the CI is breaking (fixed changes are included in master). this is nothing to do with your code. its a dependency issue that comes from cartopy. id like to merge the work that you've already contributed to this branch and release v0.6.0. we can open this pull request against 'master' once v0.6.0 is released. second, the code needs some work to be ready to publish. the docstrings need formatted properly (see other functions) so that our documentation generator can correctly parse them and create docs page from it, and there are several places with unnecessary comments (e.g., print statements that were used for debugging). also, i think we should create a new module utils/quadtree.py (or something like this) to hold some of these new functions, which should help the circular import issue that you fixed with the local import. most of this is copied from our e-mail, but wanted to share it here.

File: tests/test_spatial.py [hunk 5]
```diff
@@ -321,6 +322,40 @@ def test_geographical_area_from_bounds():
...
+    mbins = numpy.array([1])
+
+    qk = ['0', '1', '2', '3']
+    r = QuadtreeGrid2D.from_quadkeys(qk, magnitudes=mbins)
+    fcst = numpy.array([4, 4, 4, 4]).reshape(-1, 1)
+    grid_forecast = GriddedForecast(data=fcst, region=r, magnitudes=mbins)
+
+    qk_target = ['0', '10', '11', '12', '13', '2', '3']
+    r_target = QuadtreeGrid2D.from_quadkeys(qk_target, magnitudes=mbins)
+    forecast_target = forecast_mapping(r_target, grid_forecast, only_deaggregate=True)
```
Comment (wsavran) [inline tests/test_spatial.py:353]: `forecast_mapping` -> `aggregate_forecast` 
`forecast_mapping` -> `deaggregate_forecast`

File: csep/core/regions.py [hunk 6]
```diff
@@ -1331,3 +1341,313 @@ def california_quadtree_region(magnitudes=None, name="california-quadtree"):
...
+        Target grid is high-resolution at every level than the other grid.
+    """
+    from csep.core.forecasts import GriddedForecast
+    if only_deaggregate:
+        qk_target = target_grid.quadkeys
+        qk = forecast_gridded.region.quadkeys
+        data = forecast_gridded.data
+        data_mapped_qk = forecast_deaggregate_qk(qk_target, qk, data)
+        target_forecast = GriddedForecast(data=data_mapped_qk, region=target_grid,
+                                          magnitudes=forecast_gridded.magnitudes)
```
Comment (wsavran) [inline csep/core/regions.py:1644]: this should be put into its own function to specifically work with quadtrees

File: csep/core/regions.py [hunk 7]
```diff
@@ -1331,3 +1341,313 @@ def california_quadtree_region(magnitudes=None, name="california-quadtree"):
...
+            f_old = forecast_cast[i]
+            forecast_cast[i] = f_old * (area_0 / area)
+            forecast_cast.insert(i + 1, f_old * (area_1 / area))
+            forecast_cast.insert(i + 2, f_old * (area_2 / area))
+            forecast_cast.insert(i + 3, f_old * (area_3 / area))
+
+    return  numpy.vstack(forecast_cast) #numpy.array(qk_cast),
+
+
+def forecast_mapping(target_grid, forecast_gridded, only_deaggregate=False,  ncpu=None):
```
Comment (wsavran) [inline csep/core/regions.py:1623]: this should go in forecasts.py

File: csep/core/regions.py (hunk 6 above)
Comment (wsavran) [reply csep/core/regions.py:1644]: something like utils/quadtree.py

**Pull request closed without merging**